tools/schemapi/utils.py

```diff
@@ -373,140 +387,163 @@ def title(self) -> str:
-                    # In these cases, a value parameter is also always accepted.
-                    # It would be quite complex to further differentiate
-                    # between a value and a selection parameter based on
-                    # the type system (one could
-                    # try to check for the type of the Parameter.param attribute
-                    # but then we would need to write some overload signatures for
-                    # api.param).
-
-                type_representations.extend(class_names)
-            else:
-                # use RST syntax for generated sphinx docs
-                type_representations.append(rst_syntax_for_class(self.title))
+            if target == "annotation":
+                tps.update(self.title_to_type_reprs(use_concrete=use_concrete))
+            elif target == "doc":
+                tps.add(rst_syntax_for_class(self.title))
 
         if self.is_empty():
-            type_representations.append("Any")
+            tps.add("Any")
         elif self.is_literal():
             tp_str = spell_literal(self.literal)
             if for_type_hints:
                 tp_str = TypeAliasTracer.add_literal(self, tp_str, replace=True)
-            type_representations.append(tp_str)
+            tps.add(tp_str)
         elif for_type_hints and self.is_union_literal():
             it = chain.from_iterable(el.literal for el in self.anyOf)
             tp_str = TypeAliasTracer.add_literal(self, spell_literal(it), replace=True)
-            type_representations.append(tp_str)
+            tps.add(tp_str)
         elif self.is_anyOf():
-            it = (
-                s.get_python_type_representation(
-                    for_type_hints=for_type_hints, return_as_str=False
-                )
+            it_nest = (
+                s.to_type_repr(target=target, as_str=False, use_concrete=use_concrete)
                 for s in self.anyOf
             )
-            type_representations.extend(maybe_rewrap_literal(chain.from_iterable(it)))
+            tps.update(maybe_rewrap_literal(chain.from_iterable(it_nest)))
         elif isinstance(self.type, list):
             options = []
             subschema = SchemaInfo(dict(**self.schema))
             for typ_ in self.type:
+                # FIXME: Rewrite this to not mutate `SchemaInfo.schema`
                 subschema.schema["type"] = typ_
                 # We always use title if possible for nested objects
                 options.append(
-                    subschema.get_python_type_representation(
-                        for_type_hints=for_type_hints
-                    )
+                    subschema.to_type_repr(target=target, use_concrete=use_concrete)
                 )
-            type_representations.extend(options)
-        elif self.is_object() and not for_type_hints:
-            type_representations.append("dict")
+            tps.update(options)
         elif self.is_array():
```

Comment:
@binste

Related: https://github.com/vega/altair/pull/3536#discussion_r1748059188

If we can factor out this assignment, then there is a pathway to using `SchemaInfo` objects themselves in a `set`.

Provided we can find a good candidate for `SchemaInfo.__hash__()`

Reply:
I'm not sure if the benefit to this is immediately obvious, but I can explain if not?

Reply:
I think I see where you're coming from. Just creating `SchemaInfo` based on strings sounds like a simple way which works now instead of trying to rewrite that part of the code and think about hashing the `SchemaInfo` :) Except you see a very simple way, I'm not sure if it's worth the effort.

Reply:
Unrelated, but while we're chatting, I'm trying to organize a call among Altair maintainers to catch up, give an update on the NumFOCUS project summit Mattijn and I went too, and discuss priorities. Of course there is no obligation to participate and we can move important discussions to GitHub. I did not feel an email where I could invite you under so if this is of interest to you, feel free to reach out to me at binder_stefan@outlook.com so I have your contact info.

Reply:
> Unrelated, but while we're chatting, I'm trying to organize a call among Altair maintainers to catch up, give an update on the NumFOCUS project summit Mattijn and I went too, and discuss priorities. Of course there is no obligation to participate and we can move important discussions to GitHub. I did not feel an email where I could invite you under so if this is of interest to you, feel free to reach out to me at [binder_stefan@outlook.com](mailto:binder_stefan@outlook.com) so I have your contact info.

Definitely interested, just pinged you now. 

Will probably make me email public at some point, but you can share it with the other maintainers